### ISSUE TITLE
New command proposal `format number`

### ISSUE THREAD

**Issue by NotTheDr01ds** [nushell/nushell#11327]
### Related problem

On one hand, I'm half embarrassed at how long it too me to find the functionality of `into string --decimals` -- I ended up writing my own function for it, and only found `into string --decimal` a few days later by accident.  On the other hand, it's not very discoverable or (IMHO) intuitive for some formatting functionality to live in conversion commands, which is a discussion that I think has come up in the past with some other type formatting/conversion commands.

### Describe the solution you'd like

`For consistency, flexibility, and discoverability, create a new command:

```Nu
format number
```

And move the `--decimals` argument there.

At the same time, add additional functionality/formatting, such as:

`--separator`:
```Nu
> # Use locale (e.g., US English)
> 10000000 | format number --separator
10,000,000

> # Use system locale (e.g., French Canadian, IIUC?)
> 10000000 | format number --separator
10 000 000

> # Specify separator character
>1234.56 | format number --separator '_'
10_000_000

> # Separator + decimal, use locale (US English)
> 123456.78 | format number --decimals 4 --separator
123,456.7800
```

As with `format date`, the output type is `string`.

Wishlist functionality for `format number` -- [Excel/C# style formatting](https://lib.rs/crates/formato).  

### Describe alternatives you've considered

If `format number` isn't desirable, then at the least the search term `decimal` should be added to the help for `into string` to make it more discoverable.

**Comment by fdncred** [comment]
I'd probably support `format number`. Seems like a nushell-y command. 

We have something that does this somewhere because if you do `$env.config.filesize.format = b` you'll see your filesize bytes with thousands separators. Also, after you set the filesize to `b`, I think `1000 | into filesize` reports it like this too now.

It seems that you can also do `10000 | format filesize b` to produce similar results.

**Comment by amtoine** [comment]
@NotTheDr01ds 
having fun with your idea... :smirk: 
```nushell
def "format number" [
    --separator: string = ',', --decimals: int = 0
]: [int -> string, float -> string] {
    let x = $in | into string | split row '.'
    let x = [
        (
            $x.0
                | split chars
                | reverse
                | window 3 --stride 3 --remainder
                | reverse
                | each { reverse | str join }
                | str join $separator
        ),
        ($x.1? | default 0  | fill --alignment left --character "0" --width $decimals)
    ]

    $x | str join '.'
}
```
it just does not support locales but i think it does what you wanted, right? :yum:

**Comment by NotTheDr01ds** [comment]
@amtoine  Nice! I definitely learned a few things looking at this implementation.  And almost -- It's closer than I got with my simple attempt, but it still has the following issues (nothing that couldn't be fixed in Nu script, other than the locale issue):

* It only pads decimal places, not truncates:

  ```Nu
  > 1234.56789 | format number --decimals 2
  1,234.56789
  
  > 1234.56789 | into string --decimals 2
  1234.57
  ```

* `--decimals` (default value), `--decimals 0`, and `--decimals 1` are all equivalent.  As a result, it *always* outputs a decimal representation, even when an integer was provided:

  ```Nu
  > 1234 | format number
  1,234.0

  > 1234 | format number --decimals 0
  1,234.0

  > 1234 | format number --decimals 1
  1,234.0

  > 1234 | format number --decimals 2
  1,234.00
  ```

Still, I was only handling integer separators in my simplified version, and I still had a bug that I didn't notice until I looked at yours.

(Because I didn't know about `window`), I did use `group 3` instead in mine, at the cost of an extra `reverse`.  E.g.:

```Nu
12345
    | into string
    | split chars
    | reverse
    | group 3
    | each { reverse | str join }
    | reverse
    | str join ","
```

Not sure which is preferable.

I still think this would be a useful built-in, for the reasons mentioned above and more:

* `into string --decimals` is hard to find (easily fixable)
* `into string --decimals` is inconsistent with previous decisions to move formatting out of conversion commands
* `format number` could add the separator functionality
* And locale handling is important for portability.  A simplified custom command like this probably won't be robust enough to handle input like `1_000,25`, where the comma is used to delineate the decimal portion.  And I'm assuming Nushell's built-ins use Rust system locale support?

**Comment by fdncred** [comment]
I still think `format number` is a good idea but i'm worried about having a bunch of separate implementations that do similar things. I'm wondering if it would be possible to homogenize this code so that the same code could be used for `format filesize`, `into filesize`, `into string --decimals`, and `format number`?

**Comment by NotTheDr01ds** [comment]
@fdncred What are you thinking?  Combining into one command?  Or something else?

If you move this to a separate command, I'd propose deprecating `into string --decimals`, just like (we thought) `into datetime --format` was being replaced with `format date` (before we realized that `--format` was the *input* format there, not *output*).

For the others, I think having separate `format <subcommand>` follows a standard Nu pattern set by `to <subcommand_type>`, `into <subcommand_type>`, `from ...`, and others.

**Comment by fdncred** [comment]
> What are you thinking?

Really just that I'm not sure having 3 separate implementations for doing the same thing is a good design. Maybe it could be abstracted in some way and put in a separate file where every command that needs to do this type of thing and all just do it the same way.

> I'd propose deprecating ...

I agree. I like the family of `format` commands better. It's definitely more nushell-y.

**Comment by amtoine** [comment]
yeah, moving formatting bits from `into ...` to `format ...` sounds more than legit :+1:

**Comment by 132ikl** [comment]
just found this, I recently put up a PR to rename `fmt` to `format number` (#14875). but i would definitely like to see more ways to format numbers as you suggest

**Comment by sholderbach** [comment]
Yeah I think we should evolve the former `fmt` which dumps all versions eagerly into a more on demand version like @NotTheDr01ds proposes here.

**Comment by NotTheDr01ds** [comment]
We have `format bits`, but it could [use some work](https://github.com/nushell/nushell/issues/14267) to add endianness, at the least.
